packages/mdast-util-split/src/index.js

```diff
@@ -0,0 +1,146 @@
+import visit from 'unist-util-visit'
+
+const ROOT_HEADING_DEPTH = 1
+
+module.exports = function split (tree, config) {
+  const {
+    splitDepth = ROOT_HEADING_DEPTH,
+    introductionAsProperty = true,
+    conclusionAsProperty = false,
+  } = config
+
+  const firstHeading = find(tree, {type: 'heading', depth: ROOT_HEADING_DEPTH})
+  if (!firstHeading) {
+    throw new Error('No heading')
+  }
+  return splitAtDepth(
+    tree,
+    ROOT_HEADING_DEPTH,
+    {splitDepth, introductionAsProperty, conclusionAsProperty}
+  )
+}
+
+function splitAtDepth (tree, currentDepth, config) {
+  const {
+    splitDepth = ROOT_HEADING_DEPTH,
+    introductionAsProperty = true,
+    conclusionAsProperty = false,
+  } = config
+
+  const splitter = new Splitter(currentDepth)
+  const hasHeading = !!find(tree, {type: 'heading', depth: currentDepth})
+  if (!hasHeading) {
+    return tree
+  }
+  visit(tree, null, (node, index, parent) => splitter.visit(node, index, parent))
+  if (conclusionAsProperty) {
+    splitter.extractConclusions()
+  }
+  if (introductionAsProperty) {
+    splitter.extractIntroductions()
+  }
+  if (currentDepth < splitDepth) {
+    for (let i = 0; i < splitter.subTrees.length; i++) {
+      const newTree = newRootTree(splitter.subTrees[i].children.children)
+      const result = splitAtDepth(newTree, currentDepth + 1, config)
+      splitter.subTrees[i].children = result
+    }
+  }
+  return {
+    introduction: splitter.introduction,
+    trees: splitter.subTrees,
+  }
+}
+
+function find (tree, {type = 'heading', depth = ROOT_HEADING_DEPTH}) {
+  for (let i = 0; i < tree.children.length; i++) {
+    if (tree.children[i].type === type && tree.children[i].depth === depth) {
+      return tree.children[i]
+    }
+  }
+  return null
+}
+
+function newRootTree (children = []) {
+  return {
+    type: 'root',
+    children,
+  }
+}
+
+class Splitter {
+  constructor (depth = 1) {
+    this.lastIndex = -1
+    this.subTrees = []
+    this.depth = depth
+    this._introduction = newRootTree()
+  }
+  set introduction (nodes) {
```

Comment:
Using a getter/setter here makes it harder to understand IMO but I might be failing to see the benefits it brings.

`_introduction` is already initialized in the constructor, and later on (for instance l.120) when we assign to `introduction` it's hard to tell whether it enter the setter or not. I would refactor to not use these.

Reply:
I think I wanted to use a "simple" get/set at the beginning, failed and instead of reworking the whole code kept the get/set. if you want to remove this, you clearly can, it's an antipattern now.

Reply:
Good, go ahead then